File: src/main/java/com/github/hhhzzzsss/songplayer/conversion/SongParserRegistry.java
package com.github.hhhzzzsss.songplayer.conversion;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Comparator;
import java.util.List;

public class SongParserRegistry {
    public static final SongParserRegistry instance = new SongParserRegistry();
    private SongParserRegistry() {}

    private final List<SongParser> parsers = new ArrayList<>();
    private final List<String> supportedFileExtensions = new ArrayList<>();
    private List<SongParser> sortedMIMEParserList = new ArrayList<>();
    private List<SongParser> sortedFileParserList = new ArrayList<>();

    public void registerParsers(SongParser... parsers) {
        this.parsers.addAll(List.of(parsers));

        updateSortedFileParserList();
        updateSortedMIMEParserList();

        for (SongParser parser : parsers) {
            Collection<String> extensions = parser.getFileExtensions();
            if (extensions == null) continue;
            supportedFileExtensions.addAll(extensions);
        }
    }

    public List<SongParser> getMIMEParser(String mime) {
        for (SongParser converter : parsers) {
            Collection<String> mimeTypes = converter.getMIMETypes();
            if (mimeTypes == null) continue;
            if (mimeTypes.contains(mime)) return List.of(converter);
        }

        return sortedMIMEParserList;
    }

    public List<SongParser> getExtensionParser(String extension) {
        for (SongParser parser : parsers) {
            Collection<String> fileExtensions = parser.getFileExtensions();
            if (fileExtensions == null) continue;
            if (fileExtensions.contains(extension)) return List.of(parser);
        }

        return sortedFileParserList;
    }

    private void updateSortedFileParserList() {
        List<SongParser> sorted = new ArrayList<>(parsers);
        sorted.sort(Comparator.comparingInt(a -> {
            Collection<String> extensions = a.getFileExtensions();
            if (extensions == null) return 0;
            return extensions.size();
        }));
        sortedFileParserList = sorted;
    }

    private void updateSortedMIMEParserList() {
        List<SongParser> sorted = new ArrayList<>(parsers);
        sorted.sort(Comparator.comparingInt(a -> {
            Collection<String> mimeTypes = a.getMIMETypes();
            if (mimeTypes == null) return 0;
            return mimeTypes.size();
        }));
        sortedMIMEParserList = sorted;
    }

    public boolean supportsExtension(String extension) {
        return supportedFileExtensions.contains(extension);
    }
}

File: src/main/java/com/github/hhhzzzsss/songplayer/io/SongParserRegistry.kt
package com.github.hhhzzzsss.songplayer.io

object SongParserRegistry {
    private val parsers = mutableListOf<SongParser>()
    private val supportedFileExtensions = mutableListOf<String>()
    private var sortedMIMEParserList = listOf<SongParser>()
    private var sortedFileParserList = listOf<SongParser>()

    fun registerParsers(vararg parsers: SongParser) {
        this.parsers.addAll(listOf(*parsers))

        updateSortedFileParserList()
        updateSortedMIMEParserList()

        for (parser in parsers) {
            val extensions = parser.fileExtensions ?: continue
            supportedFileExtensions.addAll(extensions)
        }
    }

    fun getMIMEParser(mime: String): List<SongParser> {
        for (parser in parsers) {
            val mimeTypes = parser.mimeTypes ?: continue
            if (mimeTypes.contains(mime)) return listOf(parser)
        }

        return sortedMIMEParserList
    }

    fun getExtensionParser(extension: String): List<SongParser> {
        for (parser in parsers) {
            val fileExtensions = parser.fileExtensions ?: continue
            if (fileExtensions.contains(extension)) return listOf(parser)
        }

        return sortedFileParserList
    }

    private fun updateSortedFileParserList() {
        val sorted = parsers.toSortedSet(Comparator.comparingInt { a ->
            val fileExtensions = a.fileExtensions ?: return@comparingInt 0
            fileExtensions.size
        }).toList()
        sortedFileParserList = sorted
    }

    private fun updateSortedMIMEParserList() {
        val sorted = parsers.toSortedSet(Comparator.comparingInt { a ->
            val mimeTypes = a.mimeTypes ?: return@comparingInt 0
            mimeTypes.size
        }).toList()
        sortedMIMEParserList = sorted
    }

    fun supportsExtension(extension: String): Boolean {
        return supportedFileExtensions.contains(extension)
    }
}
